src/create-matcher.js

```diff
@@ -65,9 +65,9 @@ export function createMatcher (routes: Array<RouteConfig>): Matcher {
         hash,
         params
       }, undefined, location)
-    } else if (typeof redirect === 'string') {
+    } else if (typeof redirect === 'string' || typeof redirect === 'function') {
       // 1. resolve relative redirect
-      const rawPath = resolveRecordPath(redirect, record)
+      const rawPath = resolveRecordPath(typeof redirect === 'function' ? redirect() : redirect, record)
```

Comment:
Should we cache `type of redirect === 'function'`? 
I didn't because I thought it was too trivial.